analyzer/item/item.go

```diff
@@ -0,0 +1,211 @@
+)
+
+// ErrEmptyBatch is returned by the item based analyzer when there
+// are no work items in the batch.
+var ErrEmptyBatch = errors.New("no items in batch")
+
+type itemBasedAnalyzer[Item any] struct {
+	maxBatchSize     uint64
+	stopOnEmptyBatch bool
+	fixedInterval    time.Duration
+	analyzerName     string
+
+	processor ItemProcessor[Item]
+
+	target            storage.TargetStorage
+	logger            *log.Logger
+	queueLengthMetric prometheus.Gauge
+}
+
+var _ analyzer.Analyzer = (*itemBasedAnalyzer[any])(nil)
+
+type ItemProcessor[Item any] interface {
+	// GetItems fetches the next batch of work items.
+	GetItems(ctx context.Context, limit uint64) ([]Item, error)
+	// ProcessItem processes a single item, retrieving all required information
+	// from source storage or out of band and committing the resulting batch
+	// of queries to target storage.
+	ProcessItem(ctx context.Context, batch *storage.QueryBatch, item Item) error
+	// QueueLength returns the number of total items in the work queue. This
+	// is currently used for observability metrics.
+	QueueLength(ctx context.Context) (int, error)
+}
+
+// NewAnalyzer returns a new item based analyzer using the provided item processor.
+//
+// If stopOnEmptyBatch is true, the analyzer will process batches of items until it
+// encounters an empty batch, at which point it will terminate and return. Likely to
+// be used in the regression tests.
+//
+// If fixedInterval is provided, the analyzer will process one batch every fixedInterval.
+// By default, the analyzer will use a backoff mechanism that will attempt to run as
+// fast as possible until encountering an error.
+func NewAnalyzer[Item any](
+	name string,
+	batchSize uint64,
+	stopOnEmptyBatch bool,
+	fixedInterval time.Duration,
+	processor ItemProcessor[Item],
+	target storage.TargetStorage,
+	logger *log.Logger,
+) (analyzer.Analyzer, error) {
+	if batchSize == 0 {
+		batchSize = defaultBatchSize
+	}
+	a := &itemBasedAnalyzer[Item]{
+		batchSize,
+		stopOnEmptyBatch,
+		fixedInterval,
+		name,
+		processor,
+		target,
+		logger,
+		prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: fmt.Sprintf("%s_queue_length", name),
+			Help: fmt.Sprintf("count of items in the work queue for the %s analyzer", name),
+		}),
+	}
+	prometheus.MustRegister(a.queueLengthMetric)
+
+	return a, nil
+}
+
+// sendQueueLength reports the current number of items in the work queue to Prometheus.
+func (a *itemBasedAnalyzer[Item]) sendQueueLengthMetric(ctx context.Context) {
+	queueLength, err := a.processor.QueueLength(ctx)
+	if err != nil {
+		a.logger.Warn("error fetching queue length", "err", err)
+	}
+	a.queueLengthMetric.Set(float64(queueLength))
+}
+
+// processBatch fetches the next batch of work items, processes them in parallel, and
+// commits the resulting state changes to the database. Note that if the analyzer fails
+// to process one or more items in a batch, the entire batch is discarded and no database
+// updates are applied.
```

Comment:
This behavior mirrors the current batch/retry logic in the existing evm analyzers. However, we likely want to update this soon/now to apply db updates from items that were processed successfully, even if other items in the batch weren't, in order to make progress. Otherwise, a single bad/malicious item could stall the analyzer.

Reply:
:+1: let's update this now

Reply:
got it - updated!